napari/_qt/qt_viewer.py

```diff
@@ -1059,7 +1069,69 @@ def dropEvent(self, event):
             else:
                 filenames.append(url.toString())
 
-        self.viewer.open(filenames, stack=bool(shift_down))
+        # if trying to open as a stack, open with any available reader
+        if shift_down:
+            self.viewer.open(filenames, stack=bool(shift_down))
+            return
+
+        for filename in filenames:
+            # get available readers for this file from all registered plugins
+            readers = get_potential_readers(filename)
+            if not readers:
+                raise ValueError(
+                    trans._(
+                        'No plugin found capable of reading {filename}.',
+                        deferred=True,
+                        filename=filename,
+                    )
+                )
+
+            _, extension = os.path.splitext(filename)
+            reader_associations = get_settings().plugins.extension2reader
+            error_message = ''
+            # if we have an existing setting for this extension
+            if extension and extension in reader_associations:
+                display_name = reader_associations[extension]
+                # if this plugin is currently registered
+                if display_name in readers:
+                    try:
+                        # try to open using the associated plugin
+                        self.viewer.open(
+                            filename,
+                            plugin=readers[display_name],
+                        )
+                        # we've opened file successfully, so move on to next file
+                        return
+                    except Exception:
+                        error_message = f"Tried to open file with {display_name}, but reading failed.\n"
+                else:
+                    error_message = f"Can't find {display_name} plugin associated with {extension} files.\n"
+
+            readerDialog = QtReaderDialog(
+                parent=self,
+                pth=filename,
+                error_message=error_message,
+                readers=readers,
+            )
+            # get plugin choice from user
+            # choice is None if file was opened or user chose cancel
+            choice = _get_reader_choice_for_file(
+                readerDialog, readers, error_message
+            )
+            if choice:
+                display_name, persist_choice = choice
+                plugin_name = readers[display_name]
+                self.viewer.open(
+                    filename,
+                    plugin=plugin_name,
+                )
+                # do we have settings to save?
+                if persist_choice:
+                    # need explicit reassignment of object for persistence
+                    get_settings().plugins.extension2reader = {
+                        **reader_associations,
+                        extension: display_name,
+                    }
```

Comment:
i wonder if this could just be a method on the dialog?

```py
# it already has readers and error_message right?
# also, it has readers and can persist itself, so all we need back is the plugin_name
plugin_name = readerDialog.prompt()
if plugin_name:
    self.viewer.open(filename, plugin=plugin_name)
```

```

Reply:
I think the saving of settings shouldn't belong to the `readerDialog` method because imo we shouldn't save the setting unless opening is successful, and I think the opening should be happening in the `dropEvent`. And as for making it a method at all, it was testing where I got stuck - I know how to provide input to the dialog if I `exec` it myself in the test, but if a function/method that I call `exec`s it then I can't give it input from the test. That being said I'm going to clean it up a bit more like what you suggested above:

```python
for filename in filenames:
    reader = self._get_or_prompt_for_preferred_reader(filename)
    self.viewer.open(filename, plugin=reader)
```